src/common/bfloat16.hpp

```diff
@@ -38,6 +38,11 @@ struct bfloat16_t {
 
     DNNL_API operator float() const;
 
+    bfloat16_t &operator+=(const float a) {
```

Comment:
Is this changes something that the code in the library will actually start using or it is mostly for the completeness of the class? If the former, could you please point me to the few places where this code will be using once promoted?

Rationale: I am afraid of the differences in accuracy between the reference code that uses this class versus the optimized code that uses assembler directly.

Reply:
@emfomenk Thanks for your comments, Evarist. I just thought this would be an obvious improvement to the class. But if you think it would introduce any problematic differences between this code and the assembler based code, I can understand that this pull request might get postponed, or rejected. 

But then I'm still interested to know if it does indeed cause such differences! I saw that this pull request affects addition by a denormal, for example:

	bfloat16_t b{ FLT_MIN };
	float denorm{ FLT_MIN/2.0f };
	b += denorm;

With the original implementation, `b += denorm` would have no effect, as the denormal would be implicitly converted to zero by `bfloat_t::bfloat_t(float)`, before adding the value. With my patch, `b += denorm` would actually make `b` a little bit bigger. Wouldn't that be preferable? Can you possibly see what the assember based code would do?

FYI, I felt triggered by this issue as I'm maintaining another `bfloat_t` implementation, that is very much based upon `dnnl::impl::bfloat16_t`, at https://github.com/biovault/biovault_bfloat16

Reply:
I think the patch itself for the class makes total sense.
I just wanted to understand the origins of it. And you explained :)

> With the original implementation, `b += denorm` would have no effect, as the denormal would be implicitly converted to zero by `bfloat_t::bfloat_t(float)`, before adding the value. With my patch, `b += denorm` would actually make b a little bit bigger. Wouldn't that be preferable?

I think it would be preferable. What I worried about is that this patch might accidentally diverge the optimized code (where we might explicitly use something like `c = bfloat16_t { (float)a + (float)(bfloat16_t)b }` ) and the reference code that uses this class, hence changing the sequence. As a result we might get failures in the testing. 

I did a quick experiment:

``` cpp
bfloat16_t b{ FLT_MIN };     
float denorm{ FLT_MIN/2.0f };
                             
bfloat16_t a = b;            
a += denorm;                 
printf("%g\n", (float)a);    
                             
a = b;                       
a = (float)a + denorm;       
printf("%g\n", (float)a);    
                             
a = b;                       
a = a + denorm;              
printf("%g\n", (float)a);    
```

And got the following results:

| | Current `master` | With this PR |
| :-- | :-- | :-- |
| AVX2 | `1.17549e-38` (!) <br> `1.76324e-38` <br> `1.76324e-38` |  `1.76324e-38` <br> `1.76324e-38` <br> `1.76324e-38`  |
| AVX512 | `1.76324e-38` <br> `1.76324e-38` <br> `1.76324e-38` | `1.76324e-38` <br> `1.76324e-38` <br> `1.76324e-38`  |

The reason current master returns different values for `+=` on AVX2 and AVX512 is because the latter uses assembler directly, while for AVX2 we use the emulation code.
So, in fact your patch:
- aligns the pre-AVX512 and AVX512 behavior, and
- aligns the behavior of `a_bf16 += b_f32` with `a_bf16 = a_bf16 + b_f32`.

Which is cool!
Let me run the extended testing that covers more systems and tests. If everything is fine then we are good to go.

Thank you for the patch!

Reply:
@emfomenk Thank you very much for this very interesting feedback! 

In the mean time I did a little bit of bench-marking, and I observed that the following `for` loop running more than twice as fast after the patch:

	for (float f = (1U << FLT_MANT_DIG); f > 0; --f)
	{
		bfloat16_t b{};
		b += f;
		ASSERT_GT(b.raw_bits_, 0);
	}

Running on Windows, Visual C++ 2019 Release build, it did more than 210 ms before, and less than 90 ms after that patch  😃

Reply:
The places where the new `operator+=(float)` would be called can be found easily by locally declaring `operator+=(float) = delete;` instead, of course. I just did so, and the compilation output for oneDNN showed me six places where `operator+=(float)` would be called:

  oneDNN\src\cpu\ref_pooling.cpp(238)
  oneDNN\src\cpu\ref_deconvolution.cpp(69)
  oneDNN\src\cpu\ref_deconvolution.cpp(86)
  oneDNN\src\cpu\ref_deconvolution.cpp(111)
  oneDNN\tests\gtests\test_gemm_common.hpp(380)
  oneDNN\tests\gtests\test_lrn_forward.cpp(85)

Might be good to know that these placed may be affected by this patch, both by better performance and slightly different (apparently also better) calculation results.

Reply:
Thanks for going ahead and checking where the `operator+=(float)` is using. I check these places, and I think the new approach (w/o down-coverting the summand to `bfloat16_t`) is indeed preferable.
The internal validation is in progress, but so far so good -- no failures caused by this patch.